### PR TITLE
Attention the err of getting container ip when use special networkPlugin

### PR DESCRIPTION
**Pull request by zhengguoyong**
Fix: #17870

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/17893)

### CODE REVIEW

**Comment by dchen1107** [comment]
@zhengguoyong Like what we discussed in #17870, could you please update your pr to fail Pod on such event? Thanks!

**Comment by dchen1107** [comment]
Leave a comment at https://github.com/kubernetes/kubernetes/issues/17870#issuecomment-161079975 related to the pr. :-P

**Comment by yujuhong** [comment]
/cc @Random-Liu since he's doing a big refactoring for this part of the code.

I am unsure about letting `inspect` returns an error in such cases. IMO, The inspection itself should not fail, but the result it returns should reflect the failed network setup. Is it possible to return an empty IP string to indicate that the network plugin is not working (i.e., or is it possible for a functioning networkplugin to return an empty IP string)?

File: pkg/kubelet/dockertools/manager.go [hunk 1]
```diff
@@ -1893,7 +1894,10 @@ func (dm *DockerManager) SyncPod(pod *api.Pod, runningPod kubecontainer.Pod, pod
 
 		// Find the pod IP after starting the infra container in order to expose
 		// it safely via the downward API without a race and be able to use podIP in kubelet-managed /etc/hosts file.
-		pod.Status.PodIP = dm.determineContainerIP(pod.Name, pod.Namespace, podInfraContainer)
+		pod.Status.PodIP, err = dm.determineContainerIP(pod.Name, pod.Namespace, podInfraContainer)
+		if err != nil {
+			return err
```
Comment (Random-Liu) [inline pkg/kubelet/dockertools/manager.go]: Do we need to kill the infra container here? Why when SetUpPod() goes wrong, we kill it; but when InspectContainer goes wrong we doesn't kill it? What about here?

**Comment by zhengguoyong** [comment]
sorry for my delay reply.
As discuss in #17870, Do you think we need  attention the error from networkPlugin.Status() (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/manager.go#L330)
If we need, this PR is still needed. 
Thanks.
@dchen1107

**Comment by yujuhong** [comment]
@zhengguoyong, as I mentioned earlier in https://github.com/kubernetes/kubernetes/pull/17893#issuecomment-161081663, could we just set an empty IP string? Failing the inspection itself might be problematic as SyncPod cannot act on it at all.

**Comment by zhengguoyong** [comment]
@yujuhong 
Yes, you are right, Failing the inspection itself might be problematic as SyncPod cannot act on it at all.   If we just set an empty ip, the result is the pod status Running but has an empty ip.  If pod has no ip, it cannot work. I think uses may puzzle that why Running pod cannot work. :)
Is it better let the pod shows Pending and next SyncPod it will inspection the ip again?

**Comment by yujuhong** [comment]
> @yujuhong 
> Yes, you are right, Failing the inspection itself might be problematic as SyncPod cannot act on it at all. If we just set an empty ip, the result is the pod status Running but has an empty ip. If pod has no ip, it cannot work. I think uses may puzzle that why Running pod cannot work. :)

@zhengguoyong, the "Running" pod state, by definition, only means that kubelet has started all containers in a pod. Even if all the containers die immediately, the pod is still in the "Running" state.

> Is it better let the pod shows Pending and next SyncPod it will inspection the ip again?

The pod would still be running if it meets the "containers have been started" definition.

File: pkg/kubelet/dockertools/manager.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/dockertools/manager.go]: Inspect failures could be transient. Unless we keep track and confirm that it consistently fails, it's probably not the best to kill it right away?

**Comment by yujuhong** [comment]
The implication of this PR is that the PLEG will keep inspecting this container every relist period (1s), which is a waste of resource. I think we can tolerate that for now, but probably want a better, long-term fix that doesn't include frequent inspection.

**Comment by yujuhong** [comment]
> The implication of this PR is that the PLEG will keep inspecting this container every relist period (1s), which is a waste of resource. I think we can tolerate that for now, but probably want a better, long-term fix that doesn't include frequent inspection.

In fact, I think the pod will stay in "pending" without any extra information because of this error. @zhengguoyong, could you test your code and verify the output of such a pod? Thanks!

**Pull request closed without merging**